sdk/iot/hub/src/az_iot_hub_client_sas.c

```diff
@@ -14,104 +14,136 @@
 #define SAS_TOKEN_SKN "skn"
 
-az_result az_iot_sas_token_get_document(
-    az_span iothub_fqdn,
-    az_span device_id,
-    int32_t expiry_time_secs,
-    az_span document,
-    az_span* out_document)
+static const az_span devices_string = AZ_SPAN_LITERAL_FROM_STR(SCOPE_DEVICES_STRING);
+static const az_span modules_string = AZ_SPAN_LITERAL_FROM_STR(SCOPE_MODULES_STRING);
+
+AZ_NODISCARD az_result az_iot_hub_client_sas_signature_get(
+    az_iot_hub_client const* client,
+    uint32_t token_expiration_epoch_time,
+    az_span signature,
+    az_span* out_signature)
 {
-  AZ_PRECONDITION_VALID_SPAN(device_id, 1, false);
-  AZ_PRECONDITION_VALID_SPAN(iothub_fqdn, 1, false);
-  AZ_PRECONDITION_VALID_SPAN(document, 0, false);
-  AZ_PRECONDITION_NOT_NULL(out_document);
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION(token_expiration_epoch_time > 0);
+  AZ_PRECONDITION_VALID_SPAN(signature, 0, false);
+  AZ_PRECONDITION_NOT_NULL(out_signature);
+
+  int32_t required_length = az_span_length(client->_internal.iot_hub_hostname)
+      + az_span_length(client->_internal.device_id) + az_span_length(devices_string)
+      + 1; // 1 is for one line feed character.
 
-  az_span devices_string = AZ_SPAN_FROM_STR(SCOPE_DEVICES_STRING);
-  int32_t required_length = az_span_length(iothub_fqdn) + az_span_length(device_id)
-      + az_span_length(devices_string) + 1;
+  if (az_span_length(client->_internal.options.module_id) > 0)
+  {
+    required_length
+        += az_span_length(client->_internal.options.module_id) + az_span_length(modules_string);
+  }
 
-  AZ_RETURN_IF_NOT_ENOUGH_CAPACITY(document, required_length);
+  AZ_RETURN_IF_NOT_ENOUGH_CAPACITY(signature, required_length);
+
+  signature = az_span_append(signature, client->_internal.iot_hub_hostname);
+  signature = az_span_append(signature, devices_string);
+  signature = az_span_append(signature, client->_internal.device_id);
+
+  if (az_span_length(client->_internal.options.module_id) > 0)
+  {
+    signature = az_span_append(signature, modules_string);
+    signature = az_span_append(signature, client->_internal.options.module_id);
+  }
 
-  *out_document = document;
+  signature = az_span_append_uint8(signature, LF);
 
-  *out_document = az_span_append(*out_document, iothub_fqdn);
-  *out_document = az_span_append(*out_document, devices_string);
-  *out_document = az_span_append(*out_document, device_id);
-  *out_document = az_span_append_uint8(*out_document, LF);
+  AZ_RETURN_IF_FAILED(az_span_append_u32toa(signature, token_expiration_epoch_time, &signature));
 
-  AZ_RETURN_IF_FAILED(az_span_append_i32toa(*out_document, expiry_time_secs, out_document));
+  *out_signature = signature;
 
   return AZ_OK;
 }
 
-az_result az_iot_sas_token_generate(
-    az_span iothub_fqdn,
-    az_span device_id,
-    az_span signature,
-    int32_t expiry_time_secs,
+AZ_NODISCARD az_result az_iot_hub_client_sas_password_get(
+    az_iot_hub_client const* client,
+    az_span base64_hmac_sha256_signature,
```

Comment:
base64_sha256_hmac_signature?

Reply:
Everywhere I search the most common name is hmac-sha256, not the other way. I rather keep it like that.